Adding CCi to M2

MichaelABurr opened the pull request:
This is work started at the M2 conference in Madison.  This is the pull request to add CCi support to M2.  This includes implementations in the M2 language, the D-language, and aring support for polynomials and matrices in the back-end.

d-torrance committed:
Add patch for building factory w/ flint 3.3.0

Matrix internals have changed

MichaelABurr committed:
Merge pull request #4 from d-torrance/intervals

Add factory patch

MichaelABurr committed:
Added the CCi type

Commit:
tostringCCi added

MichaelABurr committed:
Added conversion functions and mutable CCi

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

MichaelABurr committed:
Added CCi to front end

Commit:
changes in gmp

Commit:
Merge remote-tracking branch 'refs/remotes/origin/intervals' into intervals

Commit:
changes in gmp

MichaelABurr committed:
Fixed CCi afterprint

Commit:
changes in gmp

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

Commit:
changes in gmp

andrew-tawfeek committed:
changes in gmp1

Commit:
changes in gmp

MichaelABurr committed:
Added more to creating CCi's

MichaelABurr committed:
Merged

Commit:
changes in gmp

andrew-tawfeek committed:
Merge branch 'intervals' of https://github.com/MichaelABurr/M2 into intervals

Commit:
changes in actors

d-torrance committed:
Rename RRball/CCball -> RRb/CCb (for symmetry w/ RRi/CCi)

d-torrance committed:
Actually, let's call them RRbmutable and CCbmutable

RRb and CCb will be added later and will use GC-allocated memory.

Commit:
changes in actors

andrew-tawfeek committed:
Added toCCi functionality for ZZ, QQ, RR

andrew-tawfeek committed:
changes to toCCi

MichaelABurr committed:
Added realPart and imagPart

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

MichaelABurr committed:
Added demo of two inputs to toCCi

andrew-tawfeek committed:
Added CCi to RRi

andrew-tawfeek committed:
toCCi work

MichaelABurr committed:
More of toCCi

Commit:
multiplication implemented

Commit:
integer division by CCi implemented

d-torrance committed:
Go back to RRb/CCb naming

Rather than copying RRbmutable, CCbmutable to GC-collected memory, we'll just
register finalizers for RRbcell and CCbcell

Commit:
rationals division by CCi implemented

Commit:
RR over CCi implemented

MichaelABurr committed:
Adding interval functions for CCi

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

klee669 committed:
some trials for spanCCi

klee669 committed:
some trials for spanCCi

d-torrance committed:
Revert some accidental changes

klee669 committed:
changes in reals

d-torrance committed:
RRb at top level!

andrew-tawfeek committed:
toCCi defined for pairs of ZZ, QQ, RR, RRi

Commit:
RRi division implemented

andrew-tawfeek committed:
toCCi for two input done

Commit:
RRi division implemented

Commit:
CC division implemented

klee669 committed:
promote for CCi' implemented

klee669 committed:
promote for CCi' implemented

d-torrance committed:
Fix finalization of RRb's

d-torrance committed:
Improve printing (and precision, kind of) of RRb's

Right now, precision is always 100, which is stupid.

TODO: what should the precision be?

d-torrance committed:
Construct RRb from RRi

Commit:
divisions with CCi are implemented

d-torrance committed:
Begin adding RRb stuff to reals.m2

d-torrance committed:
Fix printing of RRb's

klee669 committed:
reals updated

klee669 committed:
reals updated

d-torrance committed:
Finish toRRb constructors

d-torrance committed:
Fix precision for RRb's

MichaelABurr committed:
Adding CCi's to reals.m2

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

d-torrance committed:
Add (ring, RRb)

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

MichaelABurr committed:
Fixed issue in division

d-torrance committed:
Add toExpr(RRb,ulong) (and use it in toRRb)

d-torrance committed:
Add addition for RRb's

MichaelABurr committed:
Added single inputs to spanRRi

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

klee669 committed:
constructor function interval supports CCi

klee669 committed:
Merge remote-tracking branch 'refs/remotes/origin/intervals' into intervals

Commit:
powers with z

Commit:
powers with ZZ

andrew-tawfeek committed:
added precision to toCCi

MichaelABurr committed:
Added spanner for spanning

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

Commit:
Merge branch 'intervals' of https://github.com/MichaelABurr/M2 into intervals

andrew-tawfeek committed:
toCCi precision

Commit:
Merge branch 'intervals' of https://github.com/MichaelABurr/M2 into intervals

andrew-tawfeek committed:
merge

andrew-tawfeek committed:
merging

d-torrance committed:
Add low-level RRb constructors from ZZ/QQ, add RRb + ZZ

andrew-tawfeek committed:
merge

Commit:
CCi to the power of ZZ

Commit:
Merge branch 'intervals' of https://github.com/MichaelABurr/M2 into intervals

Commit:
CCi to the power of ZZ implemented

klee669 committed:
span function added

Commit:
RRb plus QQ implemented

d-torrance committed:
RRb -> RRi conversion

d-torrance committed:
Export CCb to top level

Commit:
RRb plus RR implemented

Commit:
additions with RRb implemented

andrew-tawfeek committed:
midpoint for CCi

andrew-tawfeek committed:
merge

klee669 committed:
(an empty interval) tag fixed

klee669 committed:
(an empty interval) tag fixed

d-torrance committed:
Add midpoint and radius/diameter for RRb

d-torrance committed:
Fix hash for RRb

d-torrance committed:
Add realPart/imaginaryPart for CCb

d-torrance committed:
Add proper string function for CCb

Also move RRb's string function to ballarith.d for consistency

d-torrance committed:
Add hash function for CCb

d-torrance committed:
Add more CCb constructors

klee669 committed:
CCi as a raw ring element

d-torrance committed:
Use unsigned longs for precision

andrew-tawfeek committed:
midpoint of CCi

klee669 committed:
intersect function is added

Commit:
isMember for CCi

Commit:
Merge branch 'intervals' of https://github.com/MichaelABurr/M2 into intervals

klee669 committed:
revert changes in reals.m2

klee669 committed:
revert changes in reals.m2

Commit:
isSubset and isMember for CCi implemented

andrew-tawfeek committed:
merge

andrew-tawfeek committed:
merge

klee669 committed:
basic Documents are added + minor fixes

MichaelABurr committed:
Intervals aring for Anton

klee669 committed:
Documents for some CCi functions + inverse interval bug fixed

Commit:
fixed to_ring_elem

MichaelABurr committed:
Removing debugging code

MichaelABurr committed:
Merge

MichaelABurr committed:
Merge pull request #5 from MichaelABurr/intervals2

Intervals2

Commit:
new CCi from RawRingElement := (CCCi,x)

Commit:
merge

MichaelABurr committed:
Merge branch 'intervals' into intervals2

MichaelABurr committed:
Merge pull request #6 from MichaelABurr/intervals2

Fixes matrices

d-torrance committed:
Remove top-level RRb/CCb types

d-torrance committed:
Add CCi <-> CCb conversion functions

So we can use arb to compute various numerical functions of CCi's

d-torrance committed:
Add trig functions for CCi

d-torrance committed:
Use inheritance for printing inexact fields

d-torrance committed:
Add some more special functions for CCi

d-torrance committed:
Add gamma function for CCi's

d-torrance committed:
Regenerate tvalues.m2 if d directory has been updated

d-torrance committed:
Revert "Regenerate tvalues.m2 if d directory has been updated"

This reverts commit b0e6bb9d1729598b837344b8c45c101769548f94.

d-torrance committed:
CCi support for more special functions

d-torrance committed:
Simplify interval w/ one complex argument

And fix some bugs

d-torrance committed:
Pass options to span

d-torrance committed:
Update "interval" docs

d-torrance committed:
Updated "diameter" docs

d-torrance committed:
Update "midpoint" docs

d-torrance committed:
Actually fixed "interval"

d-torrance committed:
Update "interval" docs so complex example isn't empty

d-torrance committed:
Move interval, diameter, and midpoint to SeeAlso in CCi doc

As subnodes, we got a "repeated references" error

d-torrance committed:
Updates to one-argument "interval"

* Replace interval(ZZ) and interval(QQ) w/ interval(Number) -- this
  way, we also support Constant objects (e.g., "interval pi")
* Support Precision option in all cases

d-torrance committed:
Use Precision option when calling "numeric" in interval(Number)

Otherwise we'll lose precision

MichaelABurr committed:
Added a test, fixed some bugs in polynomial rings with CCi and matrices

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

d-torrance committed:
Add hyperbolic functions, exp, and log for CCi

d-torrance committed:
Add exponentiation for CCi now that we have exp & log

d-torrance committed:
Add agm for both interval types

d-torrance committed:
Add abs(CCi), sign(CCi), and sign(RRi)

d-torrance committed:
abs(CCi) should return an RRi

d-torrance committed:
Add sqrt and atan for CCi

d-torrance committed:
Fix numericInterval ii (now returns a CCi instead of an empty RRi)

d-torrance committed:
Simplify span (now a binary method)

d-torrance committed:
Update span docs

d-torrance committed:
Simplify two-argument "interval" methods

MichaelABurr committed:
Added == for CCi's

d-torrance committed:
Fix 2-argument "interval" when ii is one of the arguments

d-torrance committed:
Fix precision for 2-argument "interval"

Commit:
fixed the empty interval output when the real parts are zero

MichaelABurr committed:
Fixing precision in toCCi

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

MichaelABurr committed:
Added toCCi for CCi input

d-torrance committed:
Add lngamma(CCi)

d-torrance committed:
Add size2(CCi)

d-torrance committed:
Lots o' CCi docs

d-torrance committed:
Even more CCi docs

d-torrance committed:
Add comparison of CCi's

MichaelABurr committed:
Fixed empty interval

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

d-torrance committed:
Use CCi in log(RRi) when input is negative

d-torrance committed:
Allow logs with negative bases

d-torrance committed:
Add support for acos/asin for RRi's outside [-1, 1]

d-torrance committed:
Add some RRi/CCi unit tests

d-torrance committed:
Another tostringCCi update (fix empty intervals w/ zero real part)

d-torrance committed:
Add promoters for CC -> CCi

d-torrance committed:
Make InexactNumber' inherit from Number

InexactNumber (w/o the apostrophe) already does.  This also simplifies
promotion for inexact numbers, as when the ring has Number as an
ancestor, then we promote directly in the engine rather than promoting
chain-wise through each of the base rings.

MichaelABurr committed:
Fixed isMember and added a few promotes

MichaelABurr committed:
Merged

MichaelABurr committed:
Fixed a typo/bug converting CC to CCi

MichaelABurr committed:
Improving documentation

MichaelABurr committed:
Adding tests and improving documentation

MichaelABurr committed:
Small typo fixed in numbers

MichaelABurr committed:
Merge branch 'development' into intervals

MichaelABurr committed:
Fixed merge fail

MichaelABurr committed:
Fixed typo in documentation for CCi

MichaelABurr committed:
Fixed the typo on the typo

MichaelABurr commented:
I don't know why the test is failing (did we break something or is something bigger broken?)  Help would be appreciated.  I found the change that was causing this.  But, I'm not sure the best way to do this.

MichaelABurr committed:
Fixed test, I think

d-torrance commented on M2/Macaulay2/d/gmp-old.d:
```diff
@@ -0,0 +1,2973 @@
+--This file contains gmp declarations and elementary functions.
```
Do we need gmp-old.d?

MichaelABurr replied:
No, that's not supposed to be there.  That was an emergency backup when there was a bad push this summer.

MichaelABurr replied:
I've removed the file from the repository.

MichaelABurr committed:
Deleting old backup

MichaelABurr committed:
Reverting the change of toCC to CC

d-torrance committed:
Add some tests for CCi w/ special functions

d-torrance committed:
Fix up indentation in .d files

MichaelABurr commented:
I briefly checked over the most recent changes, and I'm OK with them.  I didn't see any drastic changes there (indenting and tests, only, I think).  I also did a check with my local copy, and I didn't see any issues.  I'm ok with merging whenever you're happy (and the review has passed).

d-torrance committed:
Remove unused RRb/CCb top-level classes

MichaelABurr committed:
Added upperLeft, lowerLeft, upperRight, lowerRight for CCi

MichaelABurr committed:
Added upperLeft, lowerLeft, upperRight, lowerRight for CCi

MichaelABurr committed:
Merge branch 'intervals' of github.com:MichaelABurr/M2 into intervals

d-torrance commented on M2/Macaulay2/m2/reals.m2:
```diff
@@ -274,6 +274,30 @@ imaginaryPart QQ :=
 
+lowerLeft ZZ :=
+lowerLeft QQ :=
+lowerLeft RR := identity
```
These could be simplified to `lowerLeft Number := identity`.  Then we could do `lowerLeft pi`, `lowerLeft infinity`, etc., too.

MichaelABurr replied:
Done

MichaelABurr committed:
Simplifying lowerLeft, etc.

d-torrance committed:
Make midpoint a method

d-torrance committed:
Add midpoint for matrices and polynomials

d-torrance committed:
Fix midpoint for tower rings

Add helper methods for the "midpoint" of a ring (which is the ring
where the midpoints of polynomials lie).

d-torrance committed:
Add promoters from RRi and CCi to quotient rings

antonleykin reviewed (commented):
I'm taken away a bit by
```
i8 : # methods regularizedBeta                                                                       
                                                                                                     
o8 = 125  
```
for a method with basic usage:
```
regularizedBeta(x, a, b)                                                                
       * Inputs:                                                                                     
           * x, a "real number",                                                                     
           * a, a "real number",                                                                     
           * b, a "real number",                                                                     
       * Outputs:                                                                                    
           * a "real number",   
```
Indeed, $125 = 5^3$, but isn't it a bit too much?

d-torrance commented:
Yeah, we could probably replace most of those with something like `regularizedBeta(InexactNumber, InexactNumber, InexactNumber)`.  I'll play around with it

d-torrance committed:
Use InexactNumber for special function methods

This way, we don't have to worry about storing them as methods for all
four different InexactNumber types.

d-torrance committed:
Use arb for log1p(CC) and expm1(CC)

Previously, we were just adding/subtracting 1 first, and thus losing
precision for small inputs.

d-torrance committed:
Add sqrt for negative RRi (returns CCi)

d-torrance commented:
This is a bit more manageable!

```m2
i1 : methods regularizedBeta

o1 = {0 => (regularizedBeta, Number, InexactNumber, InexactNumber)       }
     {1 => (regularizedBeta, InexactNumber, Number, InexactNumber)       }
     {2 => (regularizedBeta, InexactNumber, InexactNumber, Number)       }
     {3 => (regularizedBeta, Number, Number, InexactNumber)              }
     {4 => (regularizedBeta, Number, InexactNumber, Number)              }
     {5 => (regularizedBeta, InexactNumber, Number, Number)              }
     {6 => (regularizedBeta, Number, Number, Number)                     }
     {7 => (regularizedBeta, InexactNumber, InexactNumber, InexactNumber)}

o1 : NumberedVerticalList
```

d-torrance committed:
Also use InexactNumber for inverse functions from Boost

Not quite correct since they're only defined for RR, but this way the
Number methods will get defined in typicalvalues.m2.

d-torrance committed:
Fix log1p(CC) test now that we use arb

d-torrance committed:
Also use InexactNumber for inverseErf

d-torrance committed:
Don't allow midpoints in quotient rings

d-torrance committed:
Add lowerLeft, etc for matrices and polynomials

left and right commented for now as they're not methods

d-torrance committed:
Fix typo

d-torrance committed:
Make "left" and "right" methods

d-torrance committed:
Preserve grading when calling midpoint(Matrix), etc.

d-torrance committed:
Add random(RRi)

d-torrance committed:
Add random(CCi)

d-torrance committed:
Merge remote-tracking branch 'origin/development' into intervals

d-torrance committed:
Merge branch 'development' into intervals

d-torrance committed:
Adapt CCi for standalone engine

d-torrance committed:
Allow printing RRi's directly to buffer

Simplifies elem_text_out and also will help us implement CCi support.

d-torrance committed:
Update printing of CCi's for standalone engine

d-torrance committed:
Merge branch 'development' into intervals

d-torrance committed:
Add polylog for CCi's

d-torrance committed:
Merge branch 'development' into intervals

d-torrance committed:
Set up promoting from real & complex interval fields

d-torrance committed:
Merge remote-tracking branch 'origin/development' into intervals

d-torrance committed:
Add a bunch of missing lift methods for inexact numbers

d-torrance committed:
Add lift(RingElement, InexactNumber)

So lift(1_(RR[x]), RR) works

d-torrance commented:
Discussed with Anton and Mike.  Not completely vetted, but we think it looks good and want it in the release.
